### PR TITLE
First implementation of tree-sitter parser

### PR DESCRIPTION
**Pull request by WilsonZiweiWang**
What it does:
1. Create a parser using the `.wasm` file generated from [here](https://github.com/amaanq/tree-sitter-bitbake)
2. Create an analyzer that utilizes the parser to do the parsing
3. Currently, it gets the declarations from the tree generated by parsing the document and uses them to create completion items
4. It shows errors in erroneous code. This does not show any details about the error for now. I only display the error when the tree says it is an error.

This PR is greatly inspired by [bash-language-server](https://github.com/bash-lsp/bash-language-server). You can refer to it for more details.

### CODE REVIEW

**Comment by WilsonZiweiWang** [comment]
I will be working on the completions according to the list in the Redmine ticket as well as the tests while this PR is being reviewed as it is a lot to digest.

File: server/src/server.ts [hunk 1]
```diff
@@ -91,7 +99,39 @@ connection.onCompletion((textDocumentPosition: TextDocumentPositionParams): Comp
   if (documentAsText === undefined) {
     return []
   }
-  return contextHandler.getComletionItems(textDocumentPosition, documentAsText)
+
+  const word = analyzer.wordAtPointFromTextPosition({
+    ...textDocumentPosition,
```
Comment (idillon-sfl) [inline server/src/server.ts:104]: It seems the name of this variable is misleading. Maybe it should be renamed "textDocumentPositionParams" or just be destructured from the beginning?

File: server/src/server.ts [hunk 2]
```diff
@@ -91,7 +99,39 @@ connection.onCompletion((textDocumentPosition: TextDocumentPositionParams): Comp
...
+      {
+        label: symbol.name,
+        kind: symbolKindToCompletionKind(symbol.kind),
+        documentation: `${symbol.name}`
+      }
+    ))
+  }
+  
+  const allCompletions = [
+    ...contextHandler.getComletionItems(textDocumentPosition, documentAsText),
```
Comment (idillon-sfl) [inline server/src/server.ts:130]: I don't think the old "getComletionItems" works. Don't hesitate to get rid of it if it is the case

**Review by idillon-sfl — COMMENTED** [review]
I noticed eslint gives a lot of errors

File: README.md [hunk 3]
```diff
@@ -31,3 +31,6 @@ For more commands, refer to the `script` section in the root `package.json`.
 
 ### Debugging
 Press `F5` or navigate to the debug section on the left of the VS Code and select the client or server to to launch the debug client.
+
+### Tree-sitter
+This extension uses tree-sitter to parse the documents. The `.wasm` file used for creating the parser is generated from [here](https://github.com/amaanq/tree-sitter-bitbake)
```
[the remark this answers is not in the document]
Comment (WilsonZiweiWang) [reply README.md:36]: We could expect constant changes from this library. I'd rather label it as the latest release

File: server/src/server.ts (hunk 1 above)
Comment (WilsonZiweiWang) [reply server/src/server.ts:104]: yep, it should be params. The original name was there at the very beginning.

File: server/src/server.ts (hunk 2 above)
Comment (WilsonZiweiWang) [reply server/src/server.ts:130]: It actually works, when you declare a variable and type its name below, you will see two completion items. One is from my code and the other is from the `contextHandler`. 
I am planning to get rid of it later when I fully understand what it was doing.

**Comment by WilsonZiweiWang** [comment]
> I noticed eslint gives a lot of errors

I fixed most of them, but some just don't make sense to me and they are annoying to fix. It's been too strict IMO. One of them tells me to use `interface` instead of `type` where it really doesn't matter much. For the ones that can be fixed by a single click on `format document`, I am all good. But I think we need to make the eslint rules a bit more flexible.

**Review by deribaucourt — COMMENTED** [review]
LGTM. Would be nice if you could fix the eslint rules or errors.

**Comment by WilsonZiweiWang** [comment]
> LGTM. Would be nice if you could fix the eslint rules or errors.

I fixed the rest of them.

**Comment by WilsonZiweiWang** [comment]
> LGTM. Would be nice if you could fix the eslint rules or errors.

Those eslint errors should be gone.

**Pull request merged**